modules/tinymce/src/themes/silver/main/ts/ui/context/ContextToolbar.ts

```diff
@@ -0,0 +1,276 @@
+          editor.shortcuts.add('ctrl+F9', 'focus statusbar', () => Keying.focusIn(comp));
+        }),
+        AlloyEvents.runOnDetached((_comp) => {
+          editor.shortcuts.remove('ctrl+F9');
+        })
+      ])
+    ])
+  });
+
+  const getScopes: () => ScopedToolbars = Thunk.cached(() => ToolbarScopes.categorise(registryContextToolbars, (toolbarApi) => {
+    // ASSUMPTION: This should only ever show one context toolbar since it's used for context forms hence [toolbarApi]
+    const alloySpec = buildToolbar([ toolbarApi ]);
+    AlloyTriggers.emitWith(contextbar, forwardSlideEvent, {
+      forwardContents: wrapInPopDialog(alloySpec)
+    });
+  }));
+
+  const buildContextToolbarGroups = (allButtons: Record<string, ContextToolbarButtonType>, ctx: InlineContent.ContextToolbarSpec) =>
+    identifyButtons(editor, { buttons: allButtons, toolbar: ctx.items, allowToolbarGroups: false }, extras, Optional.some([ 'form:' ]));
+
+  const buildContextFormGroups = (ctx: InlineContent.ContextForm, providers: UiFactoryBackstageProviders) => ContextForm.buildInitGroups(ctx, providers);
+
+  const buildToolbar = (toolbars: Array<ContextType>): AlloySpec => {
+    const { buttons } = editor.ui.registry.getAll();
+    const scopes = getScopes();
+    const allButtons: Record<string, ContextToolbarButtonType> = { ...buttons, ...scopes.formNavigators };
+
+    // For context toolbars we don't want to use floating or sliding, so just restrict this
+    // to scrolling or wrapping (default)
+    const toolbarType = getToolbarMode(editor) === ToolbarMode.scrolling ? ToolbarMode.scrolling : ToolbarMode.default;
+
+    const initGroups = Arr.flatten(Arr.map(toolbars, (ctx) =>
+      ctx.type === 'contexttoolbar' ? buildContextToolbarGroups(allButtons, ctx) : buildContextFormGroups(ctx, sharedBackstage.providers)
+    ));
+
+    return renderToolbar({
+      type: toolbarType,
+      uid: Id.generate('context-toolbar'),
+      initGroups,
+      onEscape: Optional.none,
+      cyclicKeying: true,
+      providers: sharedBackstage.providers
+    });
+  };
+
+  const getAnchor = (position: InlineContent.ContextPosition, element: Optional<SugarElement<Element>>): AnchorSpec => {
+    const anchorage = position === 'node' ? sharedBackstage.anchors.node(element) : sharedBackstage.anchors.cursor();
+    const anchorLayout = ContextToolbarAnchor.getAnchorLayout(editor, position, isTouch(), {
+      lastPos: lastPosition.get,
+      lastElement: lastElement.get,
+      isReposition: () => Optionals.is(lastTrigger.get(), TriggerCause.Reposition),
+      bounds: lastBounds.get
+    });
+    return Merger.deepMerge(anchorage, anchorLayout);
+  };
+
+  const launchContext = (toolbarApi: Array<ContextType>, elem: Optional<Element>) => {
+    launchContextToolbar.stop();
+
+    // If a mobile context menu is open, don't launch else they'll probably overlap. For android, specifically.
+    if (isTouch() && backstage.isContextMenuOpen()) {
```

Comment:
Should we move this check into a helper function since it's used more than once?

Reply:
It was like this before... if you look at the diff I haven't touched this.

Reply:
Fixed this anyways by moving it out to a `canLaunchToolbar` function.